composer/utils/checkpoint.py

```diff
@@ -532,32 +531,55 @@ def save_checkpoint(
 
     # Sharded checkpoints get their own little folder.
     if state.fsdp_sharded_state_dict_enabled:
+        if using_torch_2():
+            # Make state dict.
+            if weights_only:
+                state_dict = {'model': state.state_dict()['model']}
+            else:
+                # Dictionary must be flat to faciliate loading optimizer params.
+                state_dict = {
+                    **state.state_dict(),
+                    'rng': reproducibility.get_rng_state(),
+                }
         assert state.sharded_ckpt_prefix_dir is not None
         save_prefix_folder = state.sharded_ckpt_prefix_dir
         # New name is now Trainer.save_folder / sharded_ckpt_prefix_dir / Trainer.save_filename
         # e.g. path/to/my/checkpoints/ep{epoch}-ba{batch}/ep{epoch}-ba{batch}-rank{rank}.pt
-        save_filepath = Path(Path(filename).parent) / Path(save_prefix_folder) / Path(Path(filename).name)
+        save_dirpath = Path(Path(filename).parent) / Path(save_prefix_folder)
         # Fill in remaining placeholders.
-        save_filename = format_name_with_dist_and_time(str(save_filepath), state.run_name, state.timestamp)
-        log.debug('Saving sharded checkpoints to %s...', save_filename)
+        save_dirpath = format_name_with_dist_and_time(str(save_dirpath), state.run_name, state.timestamp)
+        save_filename = Path(save_dirpath) / Path(_TORCH_DISTRIBUTED_CHECKPOINTS_FILENAME)
     else:
         save_filename = PartialFilePath(filename).format(state, is_deepspeed)
 
     dirname = os.path.dirname(save_filename)
     if dirname:
         os.makedirs(dirname, exist_ok=True)
 
-    # only rank 0 saves the state_dict unless state.fsdp_sharded_state_dict_enabled=True.
-    if dist.get_global_rank() == 0 or state.fsdp_sharded_state_dict_enabled:
+    # All ranks save for deepspeed.
+    if is_deepspeed:
+        log.debug('Saving deepspeed checkpoints to %s...', save_filename)
+        _save_deepspeed_model(state.deepspeed_model, save_filename)
+
+    # Sharded checkpointing for torch >=2.0 uses the torch.distributed.checkpoint module.
+    elif state.fsdp_sharded_state_dict_enabled and using_torch_2():
+        import torch.distributed.checkpoint as dist_cp
```

Comment:
this needs try except i think?

Reply:
why?

Reply:
like pyright or tests or something? but if it passes, no worries.